YARG.Core/Engine/EngineManager.FailMeter.cs

```diff
@@ -73,70 +105,93 @@ public partial class EngineContainer
+                if (!note.WasFullyHit())
                 {
                     return;
                 }
-                Happiness = Math.Clamp(Happiness - HAPPINESS_PER_NOTE_MISS, -1.5f, 1f);
-                if (_engineManager.CheckForFail())
+
+                var delta = HAPPINESS_PER_NOTE_HIT * RockMeterPreset.HitRecoveryMultiplier;
+                if (_engineManager.IsAnyStarpowerActive)
                 {
-                    OnSongFailed?.Invoke();
-                    YargLogger.LogFormatDebug("Song Fail invoked after miss by player {0} with average happiness {1}", EngineId, _engineManager.Happiness);
+                    delta *= RockMeterPreset.StarPowerEffectMultiplier;
                 }
 
-                // Send under threshold event when happiness drops from above to below
-                if (Happiness < HAPPINESS_CROWD_THRESHOLD && _previousHappiness >= HAPPINESS_CROWD_THRESHOLD)
+                AddHappiness(delta);
+            }
+
+            private void OnNoteMissed<TNote>(int index, TNote note) where TNote : Note<TNote>
+            {
+                if (!note.WasFullyMissed())
                 {
-                    OnHappinessUnderThreshold?.Invoke();
+                    return;
                 }
 
-                _previousHappiness = Happiness;
+                var delta = -1 * HAPPINESS_PER_NOTE_MISS * RockMeterPreset.MissDamageMultiplier;
+                AddHappiness(delta);
             }
 
             private void OnOverstrum()
             {
-                Happiness = Math.Clamp(Happiness - HAPPINESS_PER_OVERSTRUM, -1.5f, 1f);
-                if (_engineManager.CheckForFail())
-                {
-                    OnSongFailed?.Invoke();
-                    YargLogger.LogFormatDebug("Song Fail invoked after overstrum by player {0} with average happiness {1}", EngineId, _engineManager.Happiness);
-                }
+                var delta = -1 * HAPPINESS_PER_NOTE_MISS * RockMeterPreset.OverhitDamageMultiplier;
+                AddHappiness(delta);
             }
 
             private void OnKeysOverhit(int key) => OnOverstrum();
 
             private void OnStarpowerStatus(bool isActive) => _engineManager._starpowerCount += isActive ? 1 : -1;
 
+            private void AddHappiness(float delta)
+            {
+                Happiness = Math.Clamp(Happiness + delta, HAPPINESS_MINIMUM, 1f);
+                // Send over threshold event when happiness goes from below threshold to above
+                if (Happiness >= HAPPINESS_CROWD_THRESHOLD && _previousHappiness < HAPPINESS_CROWD_THRESHOLD)
+                {
+                    OnHappinessOverThreshold?.Invoke();
+                }
+
+                // Send over threshold event when happiness goes from below threshold to above
+                if (Happiness >= HAPPINESS_CROWD_THRESHOLD && _previousHappiness < HAPPINESS_CROWD_THRESHOLD)
+                {
+                    OnHappinessOverThreshold?.Invoke();
+                }
```

Comment:
I think you intended for one of these to be over threshold and the other under, probably the following:

```suggestion
                if (Happiness >= HAPPINESS_CROWD_THRESHOLD && _previousHappiness < HAPPINESS_CROWD_THRESHOLD)
                {
                    OnHappinessOverThreshold?.Invoke();
                }
                // Send under threshold event when happiness goes from above threshold to below
                else if (Happiness < HAPPINESS_CROWD_THRESHOLD && _previousHappiness >= HAPPINESS_CROWD_THRESHOLD)
                {
                    OnHappinessUnderThreshold?.Invoke();
                }
```